consensus/polybft/proposer_calculator_test.go

```diff
@@ -0,0 +1,510 @@
+	assert.Equal(t, proposer, metadata[1].Address)
+	// validate no changes to validator set positions
+	for i, v := range vs.Accounts() {
+		assert.Equal(t, metadata[i].Address, v.Address)
+	}
+}
+
+func TestProposerCalculator_CalculateProposer(t *testing.T) {
+	t.Parallel()
+
+	validators := newTestValidatorsWithAliases([]string{"A", "B", "C", "D", "E"}, []uint64{1, 2, 3, 4, 5})
+	metadata := validators.getPublicIdentities()
+
+	vs, err := validators.toValidatorSet()
+	require.NoError(t, err)
+
+	snapshot := NewProposerSnapshot(0, metadata)
+	pc := NewProposerCalculatorFromSnapshot(snapshot, hclog.NewNullLogger())
+
+	assert.Equal(t, int64(15), vs.totalVotingPower)
+
+	currProposerAddress, err := pc.CalcProposer(0, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[4].Address, currProposerAddress)
+
+	proposerAddressR1, err := pc.CalcProposer(1, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[3].Address, proposerAddressR1)
+
+	proposerAddressR2, err := pc.CalcProposer(2, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[2].Address, proposerAddressR2)
+
+	proposerAddressR3, err := pc.CalcProposer(3, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[1].Address, proposerAddressR3)
+
+	proposerAddressR4, err := pc.CalcProposer(4, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[4].Address, proposerAddressR4)
+
+	proposerAddressR5, err := pc.CalcProposer(5, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[3].Address, proposerAddressR5)
+
+	proposerAddressR6, err := pc.CalcProposer(6, 0)
+	require.NoError(t, err)
+	assert.Equal(t, metadata[0].Address, proposerAddressR6)
+}
+
+func TestProposerCalculator_SamePriority(t *testing.T) {
+	t.Parallel()
+
+	keys, err := bls.CreateRandomBlsKeys(5)
+	require.NoError(t, err)
+
+	// at some point priorities will be the same and bytes address will be compared
+	vs, err := NewValidatorSet([]*ValidatorMetadata{
+		{
+			BlsKey:      keys[0].PublicKey(),
+			Address:     types.Address{0x1},
+			VotingPower: 1,
+		},
+		{
+			BlsKey:      keys[1].PublicKey(),
+			Address:     types.Address{0x2},
+			VotingPower: 2,
+		},
+		{
+			BlsKey:      keys[2].PublicKey(),
+			Address:     types.Address{0x3},
+			VotingPower: 3,
+		},
+	}, hclog.NewNullLogger())
+	require.NoError(t, err)
+
+	snapshot := NewProposerSnapshot(0, vs.Accounts())
+	pc := NewProposerCalculatorFromSnapshot(snapshot, hclog.NewNullLogger())
+
+	proposerR0, err := pc.CalcProposer(0, 0)
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x3}, proposerR0)
+
+	proposerR1, err := pc.CalcProposer(1, 0)
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x2}, proposerR1)
+
+	proposerR2, err := pc.CalcProposer(2, 0)
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x1}, proposerR2)
+
+	proposerR2, err = pc.CalcProposer(2, 0) // call again same round
+	require.NoError(t, err)
+	assert.Equal(t, types.Address{0x1}, proposerR2)
+}
+
+func TestProposerCalculator_ProposerSelection1(t *testing.T) {
```

Comment:
How is this test different from `TestProposerCalculator_CalculateProposer`?

Reply:
I think that `TestProposerCalculator_ProposerSelection` is copied from `Tendermint`. Not sure about this. Sasa has to help us with this. @0xSasaPrsic

Reply:
This is poorly named test, it tests that inverse order of priorities int the array does not affect selection of proposer and doing multiple iteration where frequency of the validator with the highest voting power is higher then the ones with the lower voting power.

Reply:
Ok, we just need to add that description to the unit tests. It is important to keep this context.

Reply:
Changes applied https://github.com/0xPolygon/polygon-edge/pull/1018/commits/01cea74b365bd7a4c9bdac169d934a204b77c850